qubes/device_protocol.py

```diff
@@ -0,0 +1,943 @@
+        return self._devclass
+
+    @property
+    def category(self) -> DeviceCategory:
+        """ Immutable Device category such like: 'Mouse', 'Mass_Data' etc. """
+        return self._category
+
+    @classmethod
+    def unknown(cls) -> 'DeviceInterface':
+        """ Value for unknown device interface. """
+        return cls("?******")
+
+    def __repr__(self):
+        return self._interface_encoding
+
+    def __str__(self):
+        if self.devclass == "block":
+            return "Block device"
+        if self.devclass in ("usb", "pci"):
+            result = self._load_classes(self.devclass).get(
+                self._interface_encoding[1:], None)
+            if result is None:
+                result = self._load_classes(self.devclass).get(
+                    self._interface_encoding[1:-2] + '**', None)
+            if result is None:
+                result = self._load_classes(self.devclass).get(
+                    self._interface_encoding[1:-4] + '****', None)
+            if result is None:
+                result = f"Unclassified {self.devclass} device"
+            return result
+        if self.devclass == 'mic':
+            return "Microphone"
+        return repr(self)
+
+    @staticmethod
+    def _load_classes(bus: str):
+        """
+        List of known device classes, subclasses and programming interfaces.
+        """
+        # Syntax:
+        # C class       class_name
+        #       subclass        subclass_name           <-- single tab
+        #               prog-if  prog-if_name   <-- two tabs
+        result = {}
+        with open(f'/usr/share/hwdata/{bus}.ids',
+                  encoding='utf-8', errors='ignore') as pciids:
+            class_id = None
+            subclass_id = None
+            for line in pciids.readlines():
+                line = line.rstrip()
+                if line.startswith('\t\t') and class_id and subclass_id:
+                    (progif_id, _, progif_name) = line[2:].split(' ', 2)
+                    result[class_id + subclass_id + progif_id] = \
+                        f"{class_name}: {subclass_name} ({progif_name})"
+                elif line.startswith('\t') and class_id:
+                    (subclass_id, _, subclass_name) = line[1:].split(' ', 2)
+                    # store both prog-if specific entry and generic one
+                    result[class_id + subclass_id + '**'] = \
+                        f"{class_name}: {subclass_name}"
+                elif line.startswith('C '):
+                    (_, class_id, _, class_name) = line.split(' ', 3)
+                    result[class_id + '****'] = class_name
+                    subclass_id = None
+
+        return result
+
+
+class DeviceInfo(Device):
+    """ Holds all information about a device """
+    ALLOWED_CHARS_KEY = Device.ALLOWED_CHARS_KEY + string.punctuation + ' '
+    ALLOWED_CHARS_PARAM = Device.ALLOWED_CHARS_PARAM + string.punctuation + ' '
```

Comment:
Here I'm confused, you add `string.punctuation` again. Maybe it wasn't meant to be included in nether sets in the `Device` class?
And you add space several times too.